aten/src/ATen/native/vulkan/ops/Mm.cpp

```diff
@@ -7,94 +8,207 @@ namespace vulkan {
 namespace ops {
 namespace {
 
-Tensor addmm(
-    const Tensor& self_arg,
-    const Tensor& mat1_arg,
-    const Tensor& mat2_arg,
-    const Scalar beta,
-    const Scalar alpha) {
-  api::Context* const context = api::context();
+class LinearOpContext final : public torch::jit::CustomClassHolder {
+ public:
+  static LinearOpContext create(
+      api::Resource::Pool& pool,
+      const Tensor& weight,
+      const c10::optional<Tensor>& bias);
 
-  const Tensor self = self_arg.is_vulkan() ? self_arg : self_arg.vulkan();
-  const vTensor& v_self = convert(self);
+  using State = std::tuple<Tensor, c10::optional<Tensor>>;
 
-  const Tensor mat1 = mat1_arg.is_vulkan() ? mat1_arg : mat1_arg.vulkan();
-  const vTensor& v_mat1 = convert(mat1);
+  Tensor run(const Tensor& input, float beta, float alpha) const;
+  State unpack() const;
 
-  const Tensor mat2 = mat2_arg.is_vulkan() ? mat2_arg : mat2_arg.vulkan();
-  const vTensor& v_mat2 = convert(mat2);
+ private:
+  LinearOpContext(
+      api::Resource::Pool& pool,
+      const Tensor& weight,
+      const c10::optional<Tensor>& bias);
 
-  const auto self_sizes = self.sizes();
-  const auto mat1_sizes = mat1.sizes();
-  const auto mat2_sizes = mat2.sizes();
+ private:
+  struct {
+    vTensor v_weight;
+    vTensor v_bias;
+  } packed_;
+
+  struct {
+    Tensor weight;
+    c10::optional<Tensor> bias;
+  } unpacked_;
+};
+
+vTensor pack_weights(api::Resource::Pool& pool, const Tensor& weight_arg) {
+  if (weight_arg.is_vulkan()) {
+    return convert(weight_arg);
+  }
```

Comment:
This check is no longer needed.  weight_arg.vulkan() will not be an expensive call if weight_arg.is_vulkan() is true so you can just replace both code paths with `return convert(weight_arg.vulkan());`